### DISCUSSION TITLE
Couple TensorMechanics into Thermal-Nulcear Problem

### DISCUSSION THREAD

**Discussion by js-jixu** [arfc/moltres#223] [Q&A Reactor multiphysics] [answered]
Hi, experts. 

I want to use `TensorMechanics` module of MOOSE for thermal-nuclear-mechanical coupling. We know that fuel rod will expand when its temperature rises. When I add `TensorMechanics` into original steady thermal-nuclear coupling input file(using `InversePowerMethod`), the residual shows that `TensorMechanics` doesn't take into calculation. Do you know why this happens?

```
 0 Nonlinear |R| = 4.649184e+03
    |residual|_2 of individual variables:
                   velocity: 4647.58
                   p:        4.07466e-05
                   Tf:       8.7437e-13
                   Ts:       122.101
                   disp_x:   0
                   disp_y:   0
                   disp_z:   0
                   group1:   0.00017833
                   group2:   0.00026438
                   group3:   0.000118231
                   group4:   0.000131547
                   group5:   0.00014828
                   group6:   0.000290561
                   pre1:     3.44348e-05
                   pre2:     0.000259941
                   pre3:     0.000212722
                   pre4:     0.000492182
                   pre5:     0.000287839
                   pre6:     0.000103404

    Computing Jacobian                                                                   [  9.50 s] [  140 MB]
```

Then I removed clamping and cooling, leaving only the fuel part. I want to see if TensorMechanics can work in InversePowerMethod when there is only one fuel area. Unfortunately, still not.

```
 0 Nonlinear |R| = 2.129927e+01
    |residual|_2 of individual variables:
                     Ts:     21.2993
                     disp_x: 0
                     disp_y: 0
                     disp_z: 0
                     group1: 0.000548814
                     group2: 0.00105162
                     group3: 0.000569925
                     group4: 0.000690602
                     group5: 0.000835396
                     group6: 0.00191213
                     pre1:   7.9465e-05
                     pre2:   0.000599863
                     pre3:   0.000490897
                     pre4:   0.0011358
                     pre5:   0.000664243
                     pre6:   0.000238624
```

At first, I thought TensorMechanics could not be used in InversePowerMethod. But I found an input file which uses both TensorMechanics and InversePowerMethod. So I wonder if Moltres can only solve `temperature, velocity, pressure, groupx and prex` when solving eigenvalue problems? And Moltres will not solve other variables such as `disp_x/y/z` when solving the eigenvalue problem.

I attach some files. `steady-thm-nts.i` is long, which is a complete nuclear thermal mechanical coupling problem. The `dispx/y/z` in this input file cannot be solved in `power iteration`. `TM-IMP.i` is a input file contains TensorMechanics and InversePowerMethod and can run successfully.

[thm-nts.zip](https://github.com/arfc/moltres/files/10074039/thm-nts.zip)

**Comment by js-jixu** [comment]
I replaced the original part of the `steady-thm-nts.i` with the part about `TensorMechanics` in `TM-IMP.i`. And I found that dispxyz did not participate in eigenvalue calculation. But it works in `TM-IMP.i`. So I wonder if it is because `solving the eigenvalue of diffusion equation` is incompatible with `TensorMechanics`. But I can't understand why, because TensorMechanics and diffusion equation transient calculation are compatible very well.

**Reply by js-jixu** [reply]
I changed the `initial condition` of dispxyz, and now the residual of dispxyz in eigenvalue calculation will not be 0. I think this indicates that dispxyz has participated in the calculation

```
 Power iteration= 0

Performing automatic scaling calculation

 0 Nonlinear |R| = 4.649051e+03
    |residual|_2 of individual variables:
                   velocity: 4647.58
                   p:        4.07466e-05
                   Tf:       8.26253e-13
                   Ts:       116.937
                   disp_x:   0.0106197
                   disp_y:   0.0106197
                   disp_z:   0.000100545
                   group1:   0.00017833
                   group2:   0.00026438
                   group3:   0.000118231
                   group4:   0.000131547
                   group5:   0.00014828
                   group6:   0.000290561
                   pre1:     3.44348e-05
                   pre2:     0.000259941
                   pre3:     0.000212722
                   pre4:     0.000492182
                   pre5:     0.000287839
                   pre6:     0.000103404
```

But that doesn't solve my problem. First, the calculation results are actually related to the initial value of dispxyz. That is, I set different initial values of dispxyz then get different calculation results. What I do is an eigenvalue calculation. The final result should be independent of the initial value. Second, from the output results, the distribution of dispxyz is very abnormal. It doesn't seem like a reasonable result.

The first figure is the output of disp_x solved by the eigenvalue problem and the second figure is the reasonable results.

![831913d5ab18d4884348c92d124f458](https://user-images.githubusercontent.com/99813361/204124340-49055e74-dc72-45eb-82b1-8dc757ce76c6.png)

![f2d55e241efa0d87bc8a3ed01cad74f](https://user-images.githubusercontent.com/99813361/204124344-3b8d1076-b9b0-48de-b9cd-bece4348aa87.png)

**Comment by smpark7** [comment]
I'm not familiar with the TensorMechanics parts of your input file. Try adding `use_displaced_mesh = true` to the `Nt` and `pres` blocks.

You could also try a MultiApp solve with the TensorMechanics, temperature, and flow velocity in the main app; and neutronics in the sub app. The sub app can transfer power-normalized group fluxes to your main app for the power distribution, and the main app can transfer the temperature distribution to the sub app for temperature feedback.

**Reply by js-jixu** [reply]
Hi, Sun Myung.

Sorry for the late reply, I have been using your suggestions to complete the coupling during this time. 

> Try adding use_displaced_mesh = true to the Nt and pres blocks.

I'm afraid that we can't set `use_displaced_mesh = true` directly in `Nt and pres blocks`. When I set, it turns out that:

```
unused parameter 'Nt/use_displaced_mesh'
unused parameter 'Precursors/pres/use_displaced_mesh'
```

It seems that there is no `use_displaced_mesh` in `PrecursorAction` and `NtAction`, so I don't know how `Nt and Pre` use displaced mesh. Do you know it?

I use one APP to solve the `nuclear thermal coupling`, and then use another APP to solve the `displacement field` which is a more feasible solution. If I use one MAINAPP to solve `TensorMechanics, temperature, and flow velocity`, and another SUBAPP to solve `neutronics`, it will be difficult for me to give the reasonable initial value. If I need to use MAINAPP to calculate a temperature distribution first, I need to give it an initial power distribution, which needs to be obtained from SUBAPP. And it is certainly unreasonable to directly give a sinusoidal power distribution as the initial power distribution. So it becomes a question of chicken or egg. So I used the coupling strategy of calculating the nuclear thermal coupling first and then the displacement field.

I encountered a problem, which maybe related to the `use_displaced_mesh` function of Moltres. I used my coupling method to obtain the steady-state displacements, temperature, and neutron flux distributions. Then I input them as initial values into the transient nuclear thermal mechanical coupling program. The transient output should be the same as the steady state input, but not from the results.

The first group of Postprocessors comes from the steady-state results, and the second group of Postprocessors comes from the transient results:

```
Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | average_clad_Ts| average_cool_Tf| average_fuel_Ts| bnorm          | clad_volume    | fuel_volume    | group1diff     | group1max      | group1norm     | max_clad_Ts    | max_cool_Tf    | max_fuel_Ts    | power_sum      | powernorm      | tot_fissions   |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   1.300000e+00 |   0.000000e+00 |   0.000000e+00 |   7.192427e-01 |   0.000000e+00 |   7.804963e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   1.428488e-11 |   4.487939e-01 |
|   1.000000e+00 |   7.578347e+02 |   6.828197e+02 |   9.944056e+02 |   1.324802e+00 |   1.278812e+01 |   5.156575e+01 |   5.125863e-01 |   1.395608e-01 |   1.113120e+01 |   8.558366e+02 |   8.216006e+02 |   1.616864e+03 |   2.123000e+04 |   1.437404e-11 |   4.517143e-01 |
|   2.000000e+00 |   7.578347e+02 |   6.828197e+02 |   9.944056e+02 |   1.956690e+15 |   1.278812e+01 |   5.156575e+01 |   1.632019e+15 |   2.061268e+14 |   1.644044e+16 |   8.558366e+02 |   8.216006e+02 |   1.616864e+03 |   2.123000e+04 |   2.123000e+04 |   6.671678e+14 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+

```

```
Postprocessor Values:
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
| time           | average_clad_Ts| average_cool_Tf| average_fuel_Ts| bnorm          | clad_volume    | fuel_volume    | group1diff     | group1max      | group1norm     | max_clad_Ts    | max_cool_Tf    | max_fuel_Ts    | power_sum      | powernorm      | tot_fissions   |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   7.587927e+02 |   6.839314e+02 |   9.968323e+02 |   2.001988e+15 |   1.300371e+01 |   5.266993e+01 |   5.162288e+04 |   2.061268e+14 |   1.656035e+16 |   8.577749e+02 |   8.246038e+02 |   1.620663e+03 |   2.172150e+04 |   2.172150e+04 |   6.826134e+14 |
+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+----------------+
```

We can see that their `powernorm, tot_fissions, bnorm` are different. I think this problem maybe related to `TensorMechanics` and `use_displaced_mesh`. One notable feature is that when I get the thermal expansion in the z direction of `0.6` cm in the steady state calculation, the thermal expansion I get in the transient calculation is about `1.2` cm. 

The input card is very long, so I don't attach it (I'm sorry for taking up too much of your time and I know you are very busy). I attach the steady and transient results. We can see that the initial condition of transient simulation is as same as the steady simulation results. But there is a big difference between the results of `TimeStep0` and `TimeStep1`.

[steady_and_transient_results.zip](https://github.com/arfc/moltres/files/10148290/steady_and_transient_results.zip)

**Reply by js-jixu** [reply, marked as answer]
I think I've found the reason, which is discussed on https://github.com/idaholab/moose/discussions/22721 .

Why can't I find the `Mark as answer` button...
